examples/knn_classification.py

```diff
@@ -0,0 +1,438 @@
+        train_split_labels=train_labels,
+        output=eval_embeddings,
+        target=eval_labels,
+        distributed=world_size > 1,
+        num_classes=num_classes,
+        num_valid=num_valid,
+        K=args.k,
+    )
+
+    total_num_eval = torch.tensor(num_valid, dtype=torch.int64, device=device)
+    if world_size > 1:
+        dist.reduce(total_num_eval, dst=0, op=dist.ReduceOp.SUM)
+        dist.reduce(knn_top1, dst=0, op=dist.ReduceOp.SUM)
+
+    accuracy = 100 * knn_top1.item() / total_num_eval.item()
+
+    rank_print(f'Accuracy: {accuracy:.3f}%')
+
+
+class ResizeTransform(transforms.Transform):
+    def __init__(self, size: Iterable[int], resize_multiple: int = 1):
+        super().__init__()
+
+        self.size = size
+        self.resize_multiple = resize_multiple
+
+    def _get_nearest(self, value: int):
+        return int(round(value / self.resize_multiple) * self.resize_multiple)
+
+    def _get_params(self, flat_inputs: List[Any]) -> Dict[str, Any]:
+        height, width = transforms._utils.query_size(flat_inputs)
+
+        if len(self.size) == 1:
+            # Shortest-side mode.
+            # Resize the short dimension of the image to be the specified size,
+            # and the other dimension aspect preserving
+            min_sz = min(height, width)
+            factor = self.size[0] / min_sz
+
+            rs_height = height * factor
+            rs_width = width * factor
+            size = (rs_height, rs_width)
+        elif len(self.size) == 2:
+            # Center-crop mode (the actual crop will be done by subsequent transform)
+            in_aspect = height / width
+            out_aspect = self.size[0] / self.size[1]
+
+            # Input height varies faster than output
+            if in_aspect > out_aspect:
+                scale = self.size[1] / width
+            else:
+                scale = self.size[0] / height
+
+            rs_height = height * scale
+            rs_width = width * scale
+            size = (rs_height, rs_width)
+        else:
+            raise ValueError("Unsupported resize mode")
+
+        size = tuple(self._get_nearest(d) for d in size)
+
+        return dict(size=size)
+
+    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
+        if isinstance(inpt, Image.Image):
+            inpt = inpt.convert('RGB')
+
+        size = params['size']
+
+        return transforms.functional.resize(inpt, size=size, interpolation=transforms.InterpolationMode.BICUBIC)
+
+
+def _round_up(value, multiple: int):
+    return int(math.ceil(value / multiple))
+
+
+def _collate(samples: List[Dict[str, torch.Tensor]]):
+    images = [
+        s['image']
+        for s in samples
+    ]
+    labels = [
+        s['label']
+        for s in samples
+    ]
+
+    size_groups = defaultdict(lambda: [[],[]])
+    for im, lab in zip(images, labels):
+        grp = size_groups[im.shape]
+        grp[0].append(im)
+        grp[1].append(lab)
+
+    ret = [
```

Comment:
So the data loader returns sub-batches of different image shapes? Which allows us to do the inference in one pass for each sub-batch within the batch?

Reply:
Yep, exactly